### PR TITLE
fix(shipping): jsdom-visible select2 harness + seed ajax-select2 field value

### LINKED ISSUE #447: «Список с поиском»: поле рендерится пустым и теряет уже выбранное значение
Найдено оператором на риге 21.08.2026 при проходе по чекауту в режиме «Список с поиском». Корневая причина найдена по коду, не предположение.

## Симптом

Провайдер СДЭК или DaData, поля «Регион» и «Населённый пункт» в режиме **«Список с поиском»**:

1. Поля **пустые** — нет ни значения, ни placeholder'а.
2. Ранее выбранные значения в них не отображаются.
3. Из-за полной пустоты поле схлопывается в тонкую полоску.
4. Значение **не переживает** ни перезагрузку, ни изменение связанного поля: ввёл «Татарстан» в регионе, затем «Казань» в НП, выбрал Казань — **поле региона очистилось**.

## Причина

`woodev/shipping-method/assets/js/frontend/location-select-modes.js`, `buildSelectField()`.

Функция создаёт новый `<select>` на месте `<input>` и переносит на него `id`, `name`, `className`:

```js
select.id        = input.id;
select.name      = input.name || '';
select.className = input.className;

input.parentNode.insertBefore( select, input );
input.parentNode.removeChild( input );
```

**`input.value` не переносится, и `<option>` под текущее значение не создаётся.** В ajax-стратегии
`applyEntries()` на старте не вызывается вообще — по комментарию в коде «select2's own
`ajax.transport` drives population per keystroke — nothing to pre-fetch». То есть до первого
поиска в списке нет ни одной опции.

`location-cascade.js` тоже ничего не восстанавливает: `attachOne()` вызывает рендерер и просто
записывает виджет в реестр, без возврата значения.

Отсюда все четыре симптома одной цепочкой: опций нет → селект пуст → визуально схлопнут → присвоение
`.value` селекту без такой опции даёт `''`, поэтому обратная дозаливка региона после выбора НП
**видимо очищает поле**.

## Это, судя по всему, не только отображение

`select.name` совпадает с именем поля (`shipping_state` / `shipping_city`), а пустой `<select>`
без опций отдаёт при сериализации формы пустое значение. То есть на следующем `update_checkout`
сохранённое значение перезаписывается пустым — **потеря данных, а не только косметика.**

Эта половина выведена из кода и согласуется с наблюдением оператора (поле очистилось), но
**замером на риге не подтверждена** — при взятии карточки первым делом перехватить тело POST
`update_order_review` и убедиться.

## Что сделать

Заселить селект текущим значением до инициализации select2: создать `<option>` под значение,
которое несло поле, и пометить его выбранным. Значение поля на входе доступно (`input.value`), а
подпись для него — это ровно то, что уже лежит в поле; при необходимости запись можно поднять из
`/location/select`-состояния, но начинать надо с того, чтобы **не терять уже имеющееся**.

Тест: поле в режиме `ajax-select2` с непустым значением на входе должно после `attach()` содержать
опцию с этим значением и отдавать его при сериализации. Тест обязан падать на текущем `main`.

## Связано

- Тот же проход выявил ещё три дефекта того же режима — см. соседние карточки.

### LINKED ISSUE #450: Режим «Список с поиском» не покрыт тестами вообще: в jsdom нет select2, и весь конфиг не исполняется
Заведено 21.08.2026 после того, как ручной проход оператора по чекауту нашёл в режиме «Список с
поиском» четыре дефекта подряд, ни один из которых не поймали 1260 зелёных jest-тестов. Это не
случайность, а свойство оснастки — поэтому отдельная карточка, а не строчка в одном из фиксов.

## Механизм

`location-select-modes.js` инициализирует select2 так:

```js
function ensureSelect2() {
    if ( select2Initialized || ! $select || 'function' !== typeof $select.select2 ) {
        return;
    }
    ...
    $select.select2( config );
}
```

Проверка защитная и правильная, но в jsdom `select2` **никогда** не функция: пакета
`select2`/`selectWoo` в этом репозитории нет, он приходит из WordPress/WooCommerce в рантайме.
Это прямо записано в докблоке самого файла как осознанное решение.

Следствие: **`ensureSelect2()` в тестах всегда выходит по guard'у**, и весь объект `config` —
`width`, `ajax.transport`, отсутствующие `minimumInputLength` и `delay` — не исполняется ни разу.
Тесты ходят исключительно по ветке «select2 нет», то есть по голому `<select>`.

`tests/js/location-select-modes.test.js` и `location-cascade.test.js` покрывают регистрацию в
реестре и проводку `options.fetch` — и это работает. Но **ни один тест не проверяет ни placeholder,
ни минимальную длину запроса, ни дебаунс, ни отмену запроса, ни заселение текущего значения**,
потому что дотянуться до этого кода из jsdom нечем.

То есть у продакшн-поведения этого режима покрытие **нулевое**, а прогон при этом зелёный. Ровно тот
класс тишины, который мы чиним в других местах: гарантия выглядит как гарантия, но её нет.

## Варианты (решить при взятии, не сейчас)

1. **Фейковый select2 в тест-оснастке.** Минимальная заглушка `jQuery.fn.select2`, которая
   записывает переданный config и умеет позвать `ajax.transport`. Дёшево, и сразу делает
   проверяемыми все четыре сегодняшних дефекта. Не проверяет настоящий select2, но проверяет
   **наш контракт с ним**, а сломано было именно это.
2. **Вытащить конфиг из инициализации.** Чистая функция «дай config для стратегии» тестируется без
   всякого select2 вообще. Ещё дешевле, но не покрывает поведение транспорта во времени
   (гонку ответов).
3. **e2e на риге.** Настоящий select2, настоящие ответы — но медленно и не в CI.

Вариант 1 плюс 2 закрывают всё, что сегодня сломалось; 3 — отдельный разговор.

## Почему это стоит сделать до следующего фикса режима

Три соседние карточки чинят как раз этот код. Если оснастка останется прежней, их фиксы приедут с
тестами, которые проходят по ветке «select2 нет», то есть **не доказывают починку**.

## Связано

- Три карточки на сами дефекты, найденные тем же проходом.

### PR DESCRIPTION
**Pull request by kalbac**
## What

Closes #450 first (test harness), then #447 (the empty-field bug it makes provable).

**#450 — jsdom can't see select2, so the config was dead code.** `ensureSelect2()`'s own
defensive guard (`'function' !== typeof $select.select2`) always took the "select2 absent"
branch under jest — the whole `config` object (`ajax.transport`, `placeholder`, and the
still-unfixed `minimumInputLength`/`delay` from #449) ran zero times in any test. Did both
options the card asked for, complementary:

- `selectConfigFor(strategy, seed)` — the config-building half of `ensureSelect2()` pulled out
  into a pure function: no DOM, no jQuery, no `.select2()` call. Testable with select2 absent
  entirely (`tests/js/location-select-modes.test.js`, `describe('selectConfigFor() — pure
  config builder, no select2 required', ...)`).
- `tests/js/support/fake-select2.js` — a minimal `jQuery.fn.select2` fake that records the
  config it was handed and can drive `ajax.transport`. It reproduces selectWoo's own
  `minimumInputLength` gate (`selectWoo.full.js:3885-3902`) and its store-then-abort-if-abortable
  contract (`selectWoo.full.js:3564-3571`) — `this._request != null && 'function' === typeof
  this._request.abort` — so a future #449 fix has something that actually pins "the transport's
  return value is abortable," not just its shape.

**#447 — the field renders empty and loses its value.** `buildSelectField()` moved
`id`/`name`/`className` onto the new `<select>` but never `input.value`, and the ajax path never
populated any options at start. Now, before select2 init:

- a non-empty value gets a real, pre-selected `<option>` — select2's own documented
  ["Preselect option in AJAX Select2"](select2/select2 docs, `docs/programmatic-control/add-select-clear-items.md`)
  pattern, matching the operator's own CDEK plugin
  (`plugins-reference/woocommerce-edostavka/assets/js/frontend/city-select.js:69,90-98`) — see
  "Where I departed from the reference" below;
- an empty value **with** a `placeholder`/`data-placeholder` attribute gets the blank leading
  `<option>` select2's own docs say is required for a placeholder to render at all — verified
  against `select2/select2`'s `docs/placeholders.md` ("Single select placeholders" **and**
  "Using placeholders with AJAX" — the empty option is required in both), not assumed — and
  `config.placeholder` is set to match.

## The data-loss claim — CONFIRMED, with evidence

The brief flagged this as mine, unmeasured, to prove or disprove. **It's real.** Chain, each
link checked against source, not memory:

1. `assets/js/frontend/checkout.js:532` (WooCommerce core) reads the field with jQuery:
   `s_state = $( '#shipping_state' ).val();` — not the native `.value` getter.
2. Verified empirically (this repo's own jQuery devDependency, in jsdom): a `<select>` with
   zero options has `.value === ''` (native) but **jQuery's own `.val()` returns `null`** for a
   select-one with no selection.
3. Also verified empirically: `jQuery.param({ s_state: null })` → `"s_state="` — jQuery's own
   AJAX serialization turns `null` into an **empty string that IS present on the wire**, not an
   omitted key.
4. `includes/class-wc-ajax.php:407` (`update_order_review`, read from the vendored WooCommerce
   checkout at `D:/Projects/wordpress/woocommerce`): `'shipping_state' => isset( $_POST['s_state']
   ) ? wc_clean( wp_unslash( $_POST['s_state'] ) ) : null`. Since `s_state=` **is** present
   (step 3), this evaluates to `''`, not `null`.
5. `abstract-wc-data.php:794` (`WC_Data::set_props()`): `if ( is_null( $value ) ...) { continue;
   }` — it skips `null`, but **never** skips `''`. So `''` is written straight into
   `WC()->customer`'s session shipping address.

So: empty select → `null` from `.val()` → `''` on the wire → `''` written into the session,
clobbering whatever was there before, on every checkout refresh. Pinned as a jest test (not just
this writeup) in `location-select-modes.test.js`, the `DATA LOSS CLAIM` test — asserts
`$('#shipping_state').val()` and `.serialize()` on the FIXED select produce the real value, and
fails on `main` (see below) because `main`'s `.val()` is `null` there too.

## Placeholder — decided: yes, sourced from the DOM, not a new options field

The card lists the missing placeholder as symptom (1) but doesn't ask for it explicitly; the
brief asked me to decide with reasons. Went with yes: `location-cascade.js`'s `options` object
carries no placeholder i18n string today (that file is out of my ownership for this card), but
WooCommerce's own checkout field rendering already emits a `placeholder` attribute on these
inputs, and the CDEK reference reads exactly that (`$element.attr('placeholder') ||
$element.attr('data-placeholder') || settings.placeholder`, `city-select.js:70`) — so I read it
off the DOM the same way, no new plumbing needed. Did **not** add `allowClear`: nothing in the
symptoms asks for it, and a required shipping field offering a one-click "clear to empty" button
would reintroduce exactly the data-loss chain above.

## Where I departed from the CDEK reference, and why

The reference (`city-select.js:90-98`) appends `<option value=X>X</option>` where `X =
$element.val()` — value and label are the same string, no separate lookup. I did the identical
thing (`seededOption.value = seededOption.textContent = initialValue`) for the same reason the
research doc gives: the label the field already carries **is** the value, no fetch needed. One
departure: the reference always empties the element first (`$element.empty()`) before
conditionally appending; our `select` is freshly created and never had other options, so that
step doesn't apply. No other departures.

## A real, uncarded defect found and NOT fixed here — filed as #455

Even after this fix, a **live pick** through `ajax-select2` still breaks backwards-fill into an
ancestor field of the same mode. `applyEntries()`'s ajax results give each `<option>` a
provider-KEY value (`id: entry.key`, e.g. `dadata:tatarstan`), but `location-cascade.js`'s
`writeSilently()`/`fieldValueFor()` — which drives backwards-fill — always writes the bare
component NAME (`"Татарстан"`), never the key. Confirmed with a throwaway repro (not committed):

```
after key-based pick, select.value = "dadata:tatarstan"
after backwards-fill writeSilently(el.value = bare name), select.value = ""
```

This is exactly symptom 4 from #447 ("picked Kazan, the region field cleared") — but it survives
this PR's fix because my seeded `<option>`'s value space (bare name, matching `writeSilently`)
and `applyEntries()`'s live-pick value space (provider key) are still two different spaces. A
real fix needs either this file's key scheme changed (breaks several tests currently pinning
`option.value === entry.key`, e.g. `wires the SAME options.fetch as select2's own ajax.transport`)
or `location-cascade.js::writeSilently()` made select2-aware (owned by the parallel worker on
this branch) — a coordinated, cross-file decision, not a one-line addition to either card.
Filed as **#455**, straight to Бэклог.

## Tests — new ones, verified to fail on `main`

Verified by stashing `location-select-modes.js` alone (reverting the fix, keeping the new
tests), running, then restoring. All 9 new/changed-behavior tests fail on `main`; 2 more
(unchanged-behaviour pins) correctly still pass on `main`:

**Fail on `main`:**
- `selectConfigFor() — pure config builder, no select2 required` — all 4 (function doesn't
  exist on `main`)
- `a field with a non-empty value produces a select containing that value, selected, with NO
  select2 present at all`
- `DATA LOSS CLAIM: the seeded select serializes to its actual value, not to an empty/absent
  field...`
- `placeholder: an empty field WITH a placeholder attribute gets an empty leading <option>...`
- `a non-empty value takes priority over a placeholder attribute...`
- `the seeded option is already IN THE DOM at the moment select2() is called...`

**Pass on `main` too (pin EXISTING, unchanged behaviour on purpose):**
- `a field with an empty value produces a select with NO options...`
- `the #450 fake reproduces the minimumInputLength gate and the abort contract — pinning that
  OUR transport still returns nothing (issue #449, unfixed here on purpose)`

No existing test was weakened or deleted.

## Gates

Baseline measured fresh on this branch before any change (main moved today):

- `rm -f .phpunit.result.cache && composer test:unit` — **2566 tests, 6340 assertions, 66
  skipped** (matches the required baseline exactly). No PHP touched by this PR; re-ran after
  the JS change, same result expected (no PHP files in the diff).
- `composer phpcs` — clean, 217 files.
- `composer phpstan` — `[OK] No errors`, 217 files.
- `npm run test:js -- --roots "<rootDir>/tests/js"` (from bash) — **1278 passed** (1260 baseline
  + 27 pre-existing in this file + ... i.e. all green, zero failures, zero skips).

## Bundles

Did **not** run `npm run build`. Verified first: `woodev/shipping-method/assets/js/frontend/`
is enqueued directly as a plain script
(`class-checkout-handler.php:576-577`, `wp_enqueue_script( 'woodev-location-select-modes', ...,
'js/frontend/location-select-modes.js', ... )`), never through webpack. The only files
`package.json`'s `build` script touches are the five React admin bundles under
`woodev/assets/build/` (`license-page`, `plugins-page`, `settings-page`, `setup-wizard`,
`ui-kit-gallery`) — grepped `woodev/assets/build/` for any trace of `location-select-modes` /
`selectConfigFor` / `ensureSelect2`: none. Not affected.

## Scope note

`woodev/shipping-method/assets/js/frontend/location-cascade.js` was not touched — owned by a
parallel worker on this branch. The one place I needed a change there (#455 above) is filed as
an issue, not routed around.

Closes #450
Closes #447

### CODE REVIEW

**Comment by kalbac** [comment]
Closes the critic's NOTED gap: `tests/js/support/fake-select2.js` models selectWoo's two contract points (the `minimumInputLength` gate and the store-then-abort sequence), but nothing ever drove either through it.

**What's now driven, and how:**
- New describe block `the #450 fake — minimumInputLength gate and store-then-abort, driven directly` calls `window.jQuery(el).select2(config)` with test-authored configs (not through `attachAjaxSelect2` — production never sets `minimumInputLength` today) and then `instances[0].query(term)`, proving the fake itself gates/aborts exactly like selectWoo:
  - a query shorter than the configured minimum never reaches `ajax.transport` (asserted `null` return + transport not called)
  - a query at/above the minimum does reach it
  - a second query calls `.abort()` on the first request when the transport returns something abortable
  - a second query does not throw when the previous transport call returned nothing, or an object with no `abort()`

**Passing vs. pending on #449:**
All five of the above pass — they test the fake's own fidelity to selectWoo, independent of what production wires in. One new test is `it.skip`'d: `PENDING #449: ajax.transport should return an abortable handle...`. It calls production's real `config.ajax.transport` and asserts the return value is abort()-able. It fails today (I verified by un-skipping it locally, confirming the fail, then restoring the skip) because `selectConfigFor()`'s transport never `return`s `strategy.fetchEntries(term).then(...)` — it always returns `undefined`, so selectWoo's own adapter never has anything to abort. That's exactly #449's "requests never aborted, results flicker" defect showing through; not fixed here per the brief.

**Also (small):** split the DATA LOSS CLAIM test's one paragraph-long name (mixing both paths) into two separate tests, one per path — the ajax order-review path (`jQuery('#shipping_state').val()`, mirroring checkout.js:532) and the form-submit path (`$('#checkout').serialize()`, mirroring class-wc-checkout.php:784-789). Neither assertion changed, only the split.

**Brief accuracy:** everything in the brief held up — the BLOCKING re-read was correctly rejected (verified independently: line 677/678 pin two different paths, both correct), and the NOTED gap was real and is now closed.

**Gates:** `composer test:unit` — 2566 tests, 6340 assertions, Skipped: 66 (matches the required count). `composer phpcs` clean. `composer phpstan` — No errors. `npm run test:js -- --roots "<rootDir>/tests/js"` (from bash) — 17 suites, 1285 tests, 1 skipped (the new pending-#449 case above), all green.

**Comment by kalbac** [comment]
Закрыт как поглощённый: коммиты этого PR (jsdom-видимая оснастка select2 + засев значения в ajax-select2) вошли в #461, который смержен в `main` 23.08.2026 вместе со всем стеком #461 → #462 → #464 → #467 → #468. Отдельного мержа не требуется.
